design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,483 @@
+change automatically.
+
+We will use [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision) in K8s as the Revision object.
+
+For example, we have a following component yaml:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: frontend
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash lscpu"
+```
+
+When users apply this yaml file into the system, a `ControllerRevision` object will be created.
+
+```shell script
+$ kubectl apply -f component.yaml
+```
+
+```shell script
+$ kubectl get controllerrevisions
+NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
+default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
+
+$ kubectl get controllerrevision frontend-c8bb659c5 -o yaml
+apiVersion: apps/v1
+kind: ControllerRevision
+metadata:
+  labels:
+    controller-revision-hash: c8bb659c5
+  name: frontend-c8bb659c5
+  namespace: default
+  ownerReferences:
+  - apiVersion: core.oam.dev/v1alpha2
+    blockOwnerDeletion: true
+    controller: true
+    kind: Component
+    name: frontend
+revision: 1
+data:
+  spec:
+    workload:
+      apiVersion: core.oam.dev/v1alpha2
+      kind: ContainerizedWorkload
+      spec:
+        containers:
+        - name: my-cool-workload
+          image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+          cmd:
+          - "bash lscpu"
+```
+
+Make a change to `component.yaml`:
+
+```shell script
+           cmd:
+-            - "bash lscpu"
++            - "bash top"
+```
+
+A new `ControllerRevision` will be automatically generated:
+
+```shell script
+$ kubectl get controllerrevisions
+NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
+default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
+default       frontend-a75588698    core.oam.dev/component   2          2d14h
+
+$ kubectl get controllerrevision frontend-a75588698 -o yaml
+apiVersion: apps/v1
+kind: ControllerRevision
+metadata:
+  labels:
+    controller-revision-hash: a75588698
+  name: frontend-a75588698
+  namespace: default
+  ownerReferences:
+  - apiVersion: core.oam.dev/v1alpha2
+    blockOwnerDeletion: true
+    controller: true
+    kind: Component
+    name: frontend
+revision: 2
+data:
+  spec:
+    workload:
+      apiVersion: core.oam.dev/v1alpha2
+      kind: ContainerizedWorkload
+      spec:
+        containers:
+        - name: my-cool-workload
+          image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+          cmd:
+          - "bash top"
+```
+
+Now the basic controller revision are in place, let's talk about them in Trait and ApplicationConfiguration.
```

Comment:
The design of `ControllerRevision` is cool. I really needed a versioned component, when I was trying to resolve the [issue](https://github.com/crossplane/provider-alibaba/issues/23) :)
 
```suggestion
Now the basic controller revisions are in place, let's talk about them in Trait and ApplicationConfiguration.
```